modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/CachePartitionDefragmentationManager.java

```diff
@@ -0,0 +1,889 @@
+                            pageStoreFactory
+                        );
+
+                        if (skipAlreadyDefragmentedPartition(workDir, grpId, partId, log)) {
+                            partCtx.createMappingPageStore();
+
+                            linkMapByPart.put(partId, partCtx.createLinkMapTree(false));
+
+                            continue;
+                        }
+
+                        partCtx.createMappingPageStore();
+
+                        linkMapByPart.put(partId, partCtx.createLinkMapTree(true));
+
+                        partCtx.createPartPageStore();
+
+                        copyPartitionData(partCtx, treeIter, busyLock);
+
+                        //TODO Move inside of defragmentSinglePartition.
+                        IgniteInClosure<IgniteInternalFuture<?>> cpLsnr = fut -> {
+                            if (fut.error() == null) {
+                                PageStore oldPageStore = null;
+
+                                try {
+                                    oldPageStore = filePageStoreMgr.getStore(grpId, partId);
+                                }
+                                catch (IgniteCheckedException ignore) {
+                                }
+
+                                assert oldPageStore != null;
+
+                                if (log.isDebugEnabled()) {
+                                    log.debug(S.toString(
+                                        "Partition defragmented",
+                                        "grpId", grpId, false,
+                                        "partId", partId, false,
+                                        "oldPages", oldPageStore.pages(), false,
+                                        "newPages", partCtx.partPagesAllocated.get() + 1, false,
+                                        "mappingPages", partCtx.mappingPagesAllocated.get() + 1, false,
+                                        "pageSize", pageSize, false,
+                                        "partFile", defragmentedPartFile(workDir, partId).getName(), false,
+                                        "workDir", workDir, false
+                                    ));
+                                }
+
+                                oldPageMem.invalidate(grpId, partId);
+
+                                partCtx.partPageMemory.invalidate(grpId, partId);
+
+                                DefragmentationPageReadWriteManager pageMgr = (DefragmentationPageReadWriteManager)partCtx.partPageMemory.pageManager();
+
+                                pageMgr.pageStoreMap().removePageStore(grpId, partId); // Yes, it'll be invalid in a second.
+
+                                try {
+                                    renameTempPartitionFile(workDir, partId);
+                                }
+                                catch (IgniteCheckedException e) {
+                                    throw new IgniteException(e);
```

Comment:
What would happen with this exception? You will fail the node or it will be ignored? Will it be logged?

Reply:
You throw `IgniteCheckedException` inside `renameTempPartitionFile` than catch it here and rethrow `IgniteException`. 
Is it better to throw `IgniteException` right from renameTempPartitionFile method? Why we need checked exception here?

Reply:
Checked exceptions are a standard for Java, that's why I use them when I can.
Exception will not be ignored, node will fail later because it couldn't complete checkpoint.

Reply:
I don't think we need it here, let's throw directly `IgniteException`? This will simplify the usage.

Reply:
Ok, I'm cool with it, not a big deal